File: 62-unique-paths/unique-paths.py
class Solution:
    def uniquePath(self, i:int ,j:int ,m:int ,n:int,dp : list[list[int]]) -> int:
        if(i > m or j > n):
            return 0

        if(i==m and j==n):
            return 1

        if(dp[i][j] != -1):
            return dp[i][j]

        dx:int = [1,0]
        dy:int = [0,1]
        dp[i][j] = 0

        for k in range(0,2):
            nrow:int = i + dx[k]
            ncol:int = j + dy[k]
            dp[i][j] += self.uniquePath(nrow,ncol,m,n,dp)

        return dp[i][j]

    def uniquePaths(self,m:int, n:int) -> int:
        dp = [[-1 for i in range (n)]for j in range (m)]
        return self.uniquePath(0,0,m-1,n-1,dp)
        

File: 62-unique-paths/unique-paths.cpp
class Solution {
public:
    void uniquePath(int i,int j, int m,int n, int &ans, vector<vector<int>> &dp)
    {
        if(i > m || j > n)
        {
            return;
        }

        if(i == m && j == n)
        {
            ans++;
            return;
        }

        if(dp[i][j] != -1)
        {
            ans += dp[i][j];
            return;
        }

        int dx[] = {1,0};
        int dy[] = {0,1};
        int diffAns = 0;

        for(int k=0;k<2;k++)
        {
            int nrow = i + dx[k];
            int ncol = j + dy[k];
            int prevAns = ans;
            uniquePath(nrow,ncol,m,n,ans,dp);
            diffAns += ans - prevAns;
        }

        dp[i][j] = diffAns;
        return;
    }

    int uniquePaths(int m, int n) {
        int ans = 0;
        vector<vector<int>> dp(m+1,vector<int>(n+1,-1));
        uniquePath(0,0,m-1,n-1,ans,dp);
        return ans;
    }
}; 